### ISSUE TITLE
[Proposal] Add blocTest to bloc_test

### ISSUE THREAD

**Issue by felangel** [felangel/bloc#611]
**Is your feature request related to a problem? Please describe.**
When testing it's inconvenient to assert that a bloc emitted exactly a specific set of 
states and nothing more. Also, when expecting a state that the bloc never emits, we rely on the built-in 30s timeout which is not ideal.

**Describe the solution you'd like**
```dart
// PASSES
blocTest(
  'CounterBloc emits [0] when nothing is added',
  setUp: () => CounterBloc(),
  expect: [0],
);

// PASSES
blocTest(
  'CounterBloc emits [0, 1] when CounterEvent.increment is added',
  setUp: () => CounterBloc(),
  add: CounterEvent.increment,
  expect: [0, 1],
);

/*
FAILS IMMEDIATELY (no 30s timeout)
ERROR: Expected: [0]
  Actual: [0, 1]
   Which: longer than expected at location [1]
*/
blocTest(
  'CounterBloc emits [0] when CounterEvent.increment is added',
  setUp: () => CounterBloc(),
  add: CounterEvent.increment,
  expect: [0],
);

/*
FAILS IMMEDIATELY (no 30s timeout)
ERROR: Expected: [0, 1]
  Actual: [0]
   Which: shorter than expected at location [1]
*/
blocTest(
  'CounterBloc emits [0, 1] when no events are added',
  setUp: () => CounterBloc(),
  expect: [0, 1],
);
```

`blocTest` would ensure that the bloc emits the exact list of states without allowing subsequent states through.

**Additional context**
- Would address https://github.com/felangel/bloc/pull/564 without exposing a method which is only `visibleForTesting`.
- Would not require any changes to the core bloc api.

**Comment by audkar** [comment]
Problem description looks good. I am just don't understand what solution is to solve this.

> Would also potentially require the close API to return a `Future` rather than be `void`.

Are you thinking to change behavior of existing `close` method? Currently it disposes pending events.

**Comment by audkar** [comment]
@felangel can I provide any additional feedback?

**Comment by felangel** [comment]
@audkar yes of course. I haven’t had much time to work on this and this proposal was just my initial thought. I would love your feedback 👍

**Comment by audkar** [comment]
>  I haven’t had much time to work on this <...> I would love your feedback

Sorry bad wording. I saw GH tag `feedback wanted`, so I thought that this is blocked ticket to get additional feedback.

Problem is very well described. And it is: currently there is no accessible API which lets to ensure that all pending events were handled in Bloc. 

I have no additional feedback. Just don't understand what is proposed solution is. Is it to change current `close()` behavior from discarding pending events, to handing them even after `close` is called?

**Comment by felangel** [comment]
@audkar I'm going to try to work through several implementations over the next few days in order to get a sense for what the proposed solution will be 👍

**Comment by tenhobi** [comment]
I don't have much experience with testing, but what about an interface to `add` multiple events, e.g. list? Wouldn't it be also good?

Or how to test multiple `add` with the currently proposed interface?

**Comment by audkar** [comment]
I think this issue is starting to digress from my original problem which was:
> when expecting a state that the bloc never emits, we rely on the built-in 30s timeout which is not ideal.

This is problem when doing real Test Driven Development (writing tests firsts and a lot of them). Them means that every ~1 min I have to wait 30s until I see red test.

I still want to be able to use all existing assertions like `emitsThrough` or my own techniques to test like: 
```dart
bloc.pipe(eventsCollector)
```
dart `test` package has enough assertions for my cases.

**Comment by audkar** [comment]
This wouldn't be very elegant solution :D Adding this to thousands of tests... 

Also when test fails it should show exact reason why it failed. e.g. I am using built_value's matcher `equalsBuilt`. Which provides messages like:
```
<...>
emitted an event that was <0> instead of <3> at location ['selectedVehicleIndex']
```
Making test timeout gives instead:
```
ERROR: TimeoutException after 0:00:00.010000: Test timed out after 0 seconds.
```

**Comment by felangel** [comment]
I understand your concern and in that case we can start with an `emitsExactly` operator (as initially proposed). The reason for `blocTest` was to have a standardized way of unit testing blocs that exposes a simple/flexible api while addressing all of the aforementioned concerns. In any case, we would be able to reuse the underlying implementation 👍

**Comment by felangel** [comment]
@audkar I've been messing around with different ways to implement `emitsExactly` and they all have their limitations. I think the most straightforward way to achieve what you're asking for is via something like:

```dart
test('emits [0, 1] when CounterEvent.increment is added', () async {
  fakeAsync((async) {
    final bloc = AsyncCounterBloc();
    final states = <int>[];
    bloc.listen(states.add);
    bloc.add(AsyncCounterEvent.increment);
    async.elapse(Duration(seconds: 30));
    expect(states, [0, 1]);
  });
});
```

The problem I'm having with an `emitsExactly` is you don't have control over when the bloc is created and cannot force the bloc execution to occur within the same zone.

Having a `blocTest` would solve this because it would have full control over when the bloc is created and can both create and execute the `add` within the same fake async zone. Thoughts?

**Comment by audkar** [comment]
Thank you for your attempt.

We are not going to introduce 3rd party assertions to our tests suite. I think we are sticking with Bloc fork which lets us flush events. There are more problems like silently handled Errors. So we need that modified fork anyway.

**Comment by felangel** [comment]
Can you please share how flushing events would help with this? I'm not convinced that your original pull request solves this problem.

**Comment by audkar** [comment]
[This](https://github.com/felangel/bloc/pull/564/files#diff-aaa73976ae3917742b071df4073cc5a8R388-R402) test passes with flush :)

**Comment by felangel** [comment]
The more I think about it, the more I'd rather change the bloc's close to finish processing pending events.

```dart
@override
  @mustCallSuper
  Future<void> close() =>
      _eventSubject.close().whenComplete(_stateSubject.close);
```

Then your test would be
```dart
test('description', () async {
  final expectedStates = <AsyncState>[
    AsyncState(isLoading: false, hasError: false, isSuccess: false),
    AsyncState(isLoading: true, hasError: false, isSuccess: false),
    AsyncState(isLoading: false, hasError: false, isSuccess: true),
  ];
  final states = <AsyncState>[];
  asyncBloc.listen(states.add);

  asyncBloc.add(AsyncEvent());

  await asyncBloc.close();
  expect(states, expectedStates);
});
```

I don't think there's any reason not to, thoughts?

**Comment by audkar** [comment]
I don't have opinion on that. For tests that would be good. But also that would be breaking change. Some users might rely on fact that `close()` method currently disposes all incoming events immediately.

**Comment by felangel** [comment]
@audkar I've created #639 to continue this discussion so that this issue can be for `blocTest` related discussion 👍

**Comment by felangel** [comment]
Merged in #644 and included in `bloc_test v2.0.0` 🎉
